ocaml/xapi-guard/lib/varstored_interface.ml

```diff
@@ -102,8 +75,111 @@ let serve_forever_lwt rpc_fn path =
   Lwt_switch.add_hook (Some shutdown) cleanup ;
   Lwt.return cleanup
 
+let serve_forever_lwt_callback rpc_fn path _ req body =
+  let uri = Cohttp.Request.uri req in
+  match (Cohttp.Request.meth req, Uri.path uri) with
+  | `POST, _ ->
+      let* body = Cohttp_lwt.Body.to_string body in
+      let* response =
+        Dorpc.wrap_rpc err (fun () ->
+            let call = Xmlrpc.call_of_string body in
+            (* Do not log the request, it will contain NVRAM *)
+            D.debug "Received request on %s, method %s" path call.Rpc.name ;
+            rpc_fn call
+        )
+      in
+      let body = response |> Xmlrpc.string_of_response in
+      Cohttp_lwt_unix.Server.respond_string ~status:`OK ~body ()
+  | _, _ ->
+      let body =
+        "Not allowed"
+        |> Rpc.rpc_of_string
+        |> Rpc.failure
+        |> Xmlrpc.string_of_response
+      in
+      Cohttp_lwt_unix.Server.respond_string ~status:`Method_not_allowed ~body ()
+
+(* The TPM has 3 kinds of states *)
+type state = {
+    permall: string  (** permanent storage *)
+  ; savestate: string  (** for ACPI S3 *)
+  ; volatilestate: string  (** for snapshot/migration/etc. *)
+}
+
+let deserialize t =
+  match String.split_on_char ' ' t with
+  | [permall] ->
+      (* backwards compat with reading tpm2-00.permall *)
+      {permall; savestate= ""; volatilestate= ""}
+  | [permall; savestate; volatilestate] ->
+      {permall; savestate; volatilestate}
+  | splits ->
+      Fmt.failwith "Invalid state: too many splits %d" (List.length splits)
```

Comment:
You might want to include `t` in the message.

Reply:
't' might be secret though (the entire vTPM state)

Reply:
I've added an additional check during serialization that it doesn't contain the forbidden separator char (it is base64 encoded, so it shouldn't). This will then reject malformed state early (at serialization time before the state gets corrupted), instead of deserialization time. 
This is part of this fixup: 0dc82ffbbdb1bfe9a27fa72b2588543c5b571f3d

Reply:
Good point about not logging secrets; I was looking for something to convey the shape of the string for debugging. It's good enough as it is.